planner/core/plan_cache.go

```diff
@@ -211,18 +211,32 @@ func getPointQueryPlan(stmt *ast.Prepared, sessVars *variable.SessionVars, stmtC
 }
 
 func getGeneralPlan(sctx sessionctx.Context, isGeneralPlanCache bool, cacheKey kvcache.Key, bindSQL string,
-	is infoschema.InfoSchema, stmt *PlanCacheStmt, paramTypes []*types.FieldType) (Plan,
-	[]*types.FieldName, bool, error) {
+	is infoschema.InfoSchema, stmt *PlanCacheStmt, paramTypes []*types.FieldType) (_ Plan,
+	_ []*types.FieldName, ok bool, err error) {
 	sessVars := sctx.GetSessionVars()
 	stmtCtx := sessVars.StmtCtx
 
+	// asynchronously check privilege
+	needPrivilegeCheck := false
```

Comment:
Seems that we don't need `needPrivilegeCheck`. The only possible error is from`CheckPreparedPriv`.

Reply:
Sorry I didn't get it. `needPrivilegeCheck` is to identify if the privilege check is needed. If `getValidPlanFromCache` didn't find a plan cache, then the function don't have to wait for the privilege check before exit. We can use `exists` to replace it but I think introducing a new variable is more readable.

Reply:
> Sorry I didn't get it.

Nevery mind. I got your point.

>  If getValidPlanFromCache didn't find a plan cache, then the function don't have to wait for the privilege check before exit.

`CheckPreparedPriv` definitely needs to be waited or so. There are two points:

1. we may have a temporary go routine leak, introducing unstability in tests.
2. Will `getValidPlanFromCache` take more time compared to `CheckPreparedPriv`?
If so, waiting time of`CheckPreparedPriv` could be hided completely. Waiting is cheap as no-op. We can fire async priv check before  `getValidPlanFromCache` and check the priv just after `getValidPlanFromCache`.
If not, it can not hide much execution time of `CheckPreparedPriv`. We either need to return early or check priv. We could fire async priv check after `getValidPlanFromCache` and wait. Or we just return without checking priv.

Reply:
1. Ah that's a problem
2. I did a benchmark on a sort of older version
<img width="1496" alt="image" src="https://user-images.githubusercontent.com/31720476/185083539-1b1caba5-c102-4788-bbf4-395b85945f25.png">

Reply:
And there is another problem: if `getValidPlanFromCache` failed, we don't need to wait `CheckPreparedPriv`. But we still wait for it :(

Reply:
IMO It's not a big problem. If anyone cares about the performance here, they should first make sure plan cache is hit :)